File: bfs.c
#include    <stdio.h>
#include    <stdlib.h>
static int loopstack[1000];
static int lsp=0;

void lsout() {
    for (int i=0;i<lsp;i++)
        printf("%d_",loopstack[i]);
}

void main(int argc,char *argv[]) {
    FILE *fp;
    int  c,lf;
    if (argc!=2) {
        fprintf(stderr,"Usage: bfs <file>\n");
        exit(1);
    }
    fp=fopen(argv[1],"rt");
    if (fp==NULL) {
        fprintf(stderr,"File open error.\n");
        exit(1);
    }

	printf("\t.section .text\n");
	printf("\t.globl _start\n");
    printf("_start:\n");
    printf("\tmovabs $_my_data,%%rsi\n");
    printf("\tmov\t$1,%%edx\n");

    lsp=0;
    lf='[';
    while((c=fgetc(fp))!=EOF) {
        switch (c) {
            case '>':
                printf("\tinc\t%%rsi\n");
                break;
            case '<':
                printf("\tdec\t%%rsi\n");
                break;
            case '+':
				printf("\tincb\t(%%rsi)\n");
                break;
            case '-':
				printf("\tdecb\t(%%rsi)\n");
                break;
            case '.':
				printf("\tmov\t%%edx,%%eax\n");
				printf("\tmov\t%%edx,%%ebx\n");
				printf("\tsyscall\n");
                break;
            case ',':
				printf("\txor\t%%eax, %%eax\n");
				printf("\txor\t%%ebx, %%ebx\n");
				printf("\tsyscall\n");
                break;
            case '[':
                if (lf==']')
                    loopstack[(lsp-1)]++;
                else
                    loopstack[lsp++]=1;
                lf='[';
                printf("\tjmp LE"); lsout(); printf("\n");
                printf("LB"); lsout(); printf(":\n");
                break;
            case ']':
                if (lf==']')
                    lsp--;
                lf=']';
                printf("LE"); lsout(); printf(":\n");
	            printf("\tcmp\t%%dh,(%%rsi)\n");
                printf("\tjne\tLB"); lsout(); printf("\n");
                break;
            default:
            }
    }
    fclose(fp);
    printf("\tmov\t$1,%%eax\n");
    printf("\txor\t%%ebx,%%ebx\n");
    printf("\tint $0x80\n");
    printf(".section .bss\n");
    printf("_my_data: .zero 65536\n");
    exit(0);
}

File: bfs.go
package main

import (
    "fmt"
    "os"
    "io"
)

var (
    lf int ='['
    loopstack=[]int{}
)

func lsout(loopstack [] int) {
    for i:=0;i<len(loopstack);i++ {
        fmt.Printf("%d_",loopstack[i])
    }
}

func main() {
    if len(os.Args)!=2 {
        fmt.Println("Usage: go run bfs.go file.bf >out.s")
        return
    }
	fmt.Println("\t.section .text");
	fmt.Println("\t.globl _start");
    fmt.Println("_start:");
    fmt.Println("\tmovabs $_my_data,%rsi");
    fmt.Println("\tmov\t$1,%edx");


    file,err := os.Open(os.Args[1])
    byteData,err:=io.ReadAll(file)
    _ =err

    for idx:=0;idx<len(byteData);idx++ {
        switch byteData[idx] {
        case '>':
            fmt.Println("\tinc     %rsi")
        case '<':
            fmt.Println("\tdec     %rsi")
        case '+':
            fmt.Println("\tincb     (%rsi)")
        case '-':
            fmt.Println("\tdecb     (%rsi)")
        case '.':
			fmt.Println("\tmov\t%edx,%eax");
     		fmt.Println("\tmov\t%edx,%edi");
			fmt.Println("\tsyscall");
        case ',':
        	fmt.Println("\txor\t%eax, %eax")
		    fmt.Println("\txor\t%ebx, %ebx")
			fmt.Println("\tsyscall\n")
        case '[':
            if lf==']' {
                loopstack[len(loopstack)-1]+=1
            } else {
                loopstack = append(loopstack,1)
            }
            lf='['
            fmt.Print("\tjmp     LE")
            lsout(loopstack)
            fmt.Println("")
            fmt.Print("LB")
            lsout(loopstack)
            fmt.Println(":")
        case ']':
            if lf==']' {
                loopstack=loopstack[:len(loopstack)-1]
            }
            lf=']'
            fmt.Print("LE")
            lsout(loopstack)
            fmt.Println(":")
	        fmt.Println("\tcmp\t%dh,(%rsi)");
            fmt.Print("\tjne\tLB")
            lsout(loopstack)
            fmt.Println("")
        default:
        }
    }

    fmt.Println("\tmov\t$1,%eax");
    fmt.Println("\txor\t%ebx,%ebx");
    fmt.Println("\tint $0x80");
    fmt.Println(".section .bss");
    fmt.Println("_my_data: .zero 65536");
}

